cloudinit/config/cc_apt_configure.py

```diff
@@ -244,11 +244,24 @@ def apply_apt(cfg, cloud, target):
         )
     # GH: 4344 - stop gpg-agent/dirmgr daemons spawned by gpg key imports.
     # Daemons spawned by cloud-config.service on systemd v253 report (running)
-    subp.subp(
-        ["gpgconf", "--kill", "all"],
+    gpg_process_out, _err = subp.subp(
+        ["ps", "-o", "pid,args", "-C", "dirmngr", "-C", "gpg-agent"],
         target=target,
         capture=True,
+        rcs=[0, 1],
     )
+    gpg_pids = re.findall(r"(?P<pid>\d+).*homedir \/root", gpg_process_out)
+    if len(gpg_pids) <= 2:
+        if gpg_pids:
+            LOG.debug("Killing gpg-agent and dirmngr pids: %s", gpg_pids)
+        for gpg_pid in gpg_pids:
+            os.kill(int(gpg_pid))
+    else:
+        LOG.debug(
+            "Leaving all gpg-agent and dirmngr daemons running. Greater than 2"
```

Comment:
This is actually a failure path. If >2 are found, then `cloud-init status --wait` will last forever. Are you sure we want that? All it would take to break this logic is some other service starts a daemon prior to cloud-init. It probably fits the default case, but seems fragile.

For something more correct couldn't we do something like check for for `gpg-agent` and `dirmgr` processes inside of cloud-init cgroups and end those? That way we're not broken if some process outside of cloud-init starts an agent?

Reply:
@holmanb this is no longer a failure path because we landed f7a2c48f23d924bc9a12742983f866387a248e18 which now treats the cloud-config.service as complete when MainPID== 0 and SubState==running. So we aren't blocking indefinitely anymore on subprocesses

Reply:
@blackboxsw Right, okay got it not a failure path. But we will still leak daemons in that case, right? If that's the case think that making this new code run conditionally in the case that `gpgconf` doesn't exist on that path would be safer?

Up to you - I'll remove my blocking review.

Reply:
Also, I've dropped this else now anyway since we are more explicitly matching on processed spawned by PPID=1. This prevents cloud-init from accidentally clobbering other user's gpg-agent/dirmngr.